### ISSUE TITLE
Seemingly inconsistent result-merged.hdf5 and clusters-merged.hdf5

### ISSUE THREAD

**Issue by morales-gregorio** [spyking-circus/spyking-circus#198]
Hey there!

I am using spyking circus (pip installed version 0.8.5 in ubuntu) for my next publication and have come across some inconsistencies in my results. I am not sure if this is a mistake I am making from my parameter selection or if there might be a problem.

Let me demonstrate by showing you my python terminal commands. I first load my `result-merged.hdf5` and `clusters-merged.hdf5`:
```
>>> import h5py
>>> result = h5py.File('NSP1.result-merged.hdf5', mode='r')
>>> clusters = h5py.File('NSP1.clusters-merged.hdf5', mode='r')
```
I have 691 templates:
```
>>> len(list(clusters['electrodes']))
691
```
Just for illustration if I look at the first template:
```
>>> list(clusters['electrodes'])[0]
0
```
I can see that it corresponds to electrode `0`. I have a look at the times recorded in this file:
```
>>> clusters['times_0']
<HDF5 dataset "times_0": shape (12214,), type "<u4">
```
The problem comes when I look at the same template in `result`:
```
>>> result['spiketimes']['temp_0']
<HDF5 dataset "temp_0": shape (19426,), type "<u4">
```
There are more spikes in the template than the electrode! As a final check, I looked at how many of the times are actually the same:
```
>>> import numpy as np
>>> c0 = np.array(list(clusters['times_0']))
>>> r0 = np.array(list(result['spiketimes']['temp_0']))
>>> intersection = np.intersect1d(c0, r0)
>>> intersection.shape
(6805,)
```
*Is it normal that there are more spiketimes registered to a template than spikes where found in the same electrode?* In particular spiketimes that do not even appear on that electrode. 

Or even more worrisome: why are all the spiketimes from my template not found within the ones from the electrode?

In principle my electrodes are far enough (400 microns) that no spike should be seen in any two of them, thus I should not have any spikes from other electrodes added to them. I tried to set the parameters in such a way that this would not happen. However, from plotting the waveforms (by manually extracting them from the spiketimes) I get the feeling that this is exactly what is happening.

Here my parameters (after running spyking circus) in case you can spot the problem there.
```
### Note that depending on the file format selected, the parameters in the data section can vary
### You should refer to the documentation to know what are the needed parameters for a given
### file format. Otherwise, launch the code and a message will tell you what is needed

[data]
file_format    = numpy                 # Can be raw_binary, openephys, hdf5, ... See >> spyking-circus help -i for more info
stream_mode    = None                  # None by default. Can be multi-files, or anything depending to the file format
mapping        = probefile.prb         # Mapping of the electrode (see http://spyking-circus.rtfd.ord)
suffix         =                       # Suffix to add to generated files, if needed
overwrite      = False                 # Filter or remove artefacts on site (if write access is possible). Data are duplicated otherwise
parallel_hdf5  = True                  # Use the parallel HDF5 feature (if available)
output_dir     =                       # By default, generated data are in the same folder as the data.
blosc_compress = True
sampling_rate  = 30000.

[detection]
radius         = 1          # Radius [in um] (if auto, read from the prb file)
N_t            = 3          # Width of the templates [in ms]
spike_thresh   = 8          # Threshold for spike detection
peaks          = negative       # Can be negative (default), positive or both
alignment      = False      # Realign the waveforms by oversampling
dead_channels  =            # If not empty or specified in the probe, a dictionary }channel_group : [list_of_valid_ids]}

[filtering]
cut_off        = 250, 5000  # Min and Max (auto=nyquist) cut off frequencies for the band pass butterworth filter [Hz]
filter         = True       # If True, then a low-pass filtering is performed
remove_median  = True       # If True, median over all channels is substracted to each channels (movement artifacts)
common_ground  =            # If you want to use a particular channel as a reference ground: should be a channel number

[triggers]
trig_file      =            # External stimuli to be considered as putative artefacts [in trig units] (see documentation)
trig_windows   =            # The time windows of those external stimuli [in trig units]
trig_unit      = ms         # The unit in which times are expressed: can be ms or timestep
clean_artefact = False      # If True, external artefacts induced by triggers will be suppressed from data
dead_file      =            # Portion of the signals that should be excluded from the analysis [in dead units]
dead_unit      = ms         # The unit in which times for dead regions are expressed: can be ms or timestep
ignore_times   = False      # If True, any spike in the dead regions will be ignored by the analysis
make_plots     =            # Generate sanity plots of the averaged artefacts [Nothing or None if no plots]

[whitening]
chunk_size     = 30         # Size of the data chunks [in s]
safety_time    = auto       # Temporal zone around which templates are isolated [in ms, or auto]
spatial        = True       # Perform spatial whitening
max_elts       = 1000       # Max number of events per electrode (should be compatible with nb_elts)
nb_elts        = 0.8        # Fraction of max_elts that should be obtained per electrode [0-1]
output_dim     = 5          # Can be in percent of variance explain, or num of dimensions for PCA on waveforms

[clustering]
extraction     = median-raw # Can be either median-raw (default), median-pca, mean-pca, mean-raw
safety_space   = True       # If True, we exclude spikes in the vicinity of a selected spikes
safety_time    = auto       # Temporal zone around which templates are isolated [in ms, or auto]
max_elts       = 10000      # Max number of events per electrode (should be compatible with nb_elts)
nb_elts        = 0.8        # Fraction of max_elts that should be obtained per electrode [0-1]
nclus_min      = 0.002      # Min number of elements in a cluster (given in percentage) [0-1]
max_clusters   = 10         # Maximal number of clusters for every electrodes
nb_repeats     = 3          # Number of passes used for the clustering
smart_search   = True       # Activate the smart search mode
smart_select   = False      # Experimental: activate the smart selection of centroids (max_clusters is ignored)
merging_method = bhatta     # Method to perform local merges (distance, dip, folding, nd-folding, bhatta)
merging_param  = default    # Merging parameter (see docs) (3 if distance, 0.5 if dip, 1e-9 if folding, 2 if bhatta)
cc_merge       = 1.000      # If CC between two templates is higher, they are merged
dispersion     = (5, 5)     # Min and Max dispersion allowed for amplitudes [in MAD]
noise_thr      = 0.8        # Minimal amplitudes are such than amp*min(templates) < noise_thr*threshold in [0-1]
remove_mixture = True       # At the end of the clustering, we remove mixtures of templates
cc_mixtures    = 0.75       # If CC between a sum of two templates and a template is higher, it is removed
make_plots     = png        # Generate sanity plots of the clustering [Nothing or None if no plots]

[fitting]
chunk_size     = 1          # Size of chunks used during fitting [in second]
amp_limits     = (0.3, 5)   # Amplitudes for the templates during spike detection [if not auto]
amp_auto       = True       # True if amplitudes are adjusted automatically for every templates
max_chunk      = inf        # Fit only up to max_chunk
collect_all    = False      # If True, one garbage template per electrode is created, to store unfitted spikes

[merging]
cc_overlap     = 0.70       # Only templates with CC higher than cc_overlap may be merged
cc_bin         = 2          # Bin size for computing CC [in ms]
correct_lag    = True       # If spikes are aligned when merging. May be better for phy usage
auto_mode      = 0.2        # If >0, merging will be automatic (see doc, 0.1 is a good value)

[converting]
erase_all      = True       # If False, a prompt will ask you to export if export has already been done
export_pcs     = none       # Can be prompt [default] or in none, all, some
export_all     = False      # If True, unfitted spikes will be exported as the last Ne templates
sparse_export  = False      # For recent versions of phy, and large number of templates/channels

[validating]
nearest_elec   = auto       # Validation channel (e.g. electrode closest to the ground truth cell)
max_iter       = 200        # Maximum number of iterations of the stochastic gradient descent (SGD)
learning_rate  = 1.0e-3     # Initial learning rate which controls the step-size of the SGD
roc_sampling   = 10         # Number of points to estimate the ROC curve of the BEER estimate
test_size      = 0.3        # Portion of the dataset to include in the test split
radius_factor  = 0.5        # Radius factor to modulate physical radius during validation
juxta_dtype    = uint16     # Type of the juxtacellular data
juxta_thresh   = 6          # Threshold for juxtacellular detection
juxta_valley   = False      # True if juxta-cellular spikes are negative peaks
juxta_spikes   =            # If none, spikes are automatically detected based on juxta_thresh
filter         = True       # If the juxta channel need to be filtered or not
make_plots     = png        # Generate sanity plots of the validation [Nothing or None if no plots]

[extracting]
safety_time    = 1          # Temporal zone around which spikes are isolated [in ms]
max_elts       = 1000       # Max number of collected events per templates
output_dim     = 5          # Percentage of variance explained while performing PCA
cc_merge       = 0.975      # If CC between two templates is higher, they are merged
noise_thr      = 0.8        # Minimal amplitudes are such than amp*min(templates) < noise_thr*threshold

[noedits]
filter_done    = True              #!! AUTOMATICALLY EDITED: DO NOT MODIFY !!
artefacts_done = False      # Will become True automatically after removing artefacts
median_done    = True              #!! AUTOMATICALLY EDITED: DO NOT MODIFY !!
ground_done    = False      # Will become True automatically after removing common ground
```
And also the probefile in case you think the problem can be there:
```
total_nb_channels = 128
radius            = 100
channel_groups    = {1 : {}}


channel_groups[1]["channels"] = list(range(128))
channel_groups[1]["graph"]    = []
channel_groups[1]["geometry"] = {
 0: [0, 2800],
 1: [0, 2400],
 2: [0, 2000],
 3: [0, 1600],
 4: [0, 1200],
 5: [0, 800],
 6: [0, 400],
 7: [0, 0],
 8: [400, 2800],
 9: [400, 2400],
 10: [400, 2000],
 11: [400, 1600],
 12: [400, 1200],
 13: [400, 800],
 14: [400, 400],
 15: [400, 0],
 16: [800, 2800],
 17: [800, 2400],
 18: [800, 2000],
 19: [800, 1600],
 20: [800, 1200],
 21: [800, 800],
 22: [800, 400],
 23: [800, 0],
 24: [1200, 2800],
 25: [1200, 2400],
 26: [1200, 2000],
 27: [1200, 1600],
 28: [1200, 1200],
 29: [1200, 800],
 30: [1200, 400],
 31: [1200, 0],
 32: [1600, 2800],
 33: [1600, 2400],
 34: [1600, 2000],
 35: [1600, 1600],
 36: [1600, 1200],
 37: [1600, 800],
 38: [1600, 400],
 39: [1600, 0],
 40: [2000, 2800],
 41: [2000, 2400],
 42: [2000, 2000],
 43: [2000, 1600],
 44: [2000, 1200],
 45: [2000, 800],
 46: [2000, 400],
 47: [2000, 0],
 48: [2400, 2800],
 49: [2400, 2400],
 50: [2400, 2000],
 51: [2400, 1600],
 52: [2400, 1200],
 53: [2400, 800],
 54: [2400, 400],
 55: [2400, 0],
 56: [2800, 2800],
 57: [2800, 2400],
 58: [2800, 2000],
 59: [2800, 1600],
 60: [2800, 1200],
 61: [2800, 800],
 62: [2800, 400],
 63: [2800, 0],
 64: [50000, 2800],
 65: [50000, 2400],
 66: [50000, 2000],
 67: [50000, 1600],
 68: [50000, 1200],
 69: [50000, 800],
 70: [50000, 400],
 71: [50000, 0],
 72: [50400, 2800],
 73: [50400, 2400],
 74: [50400, 2000],
 75: [50400, 1600],
 76: [50400, 1200],
 77: [50400, 800],
 78: [50400, 400],
 79: [50400, 0],
 80: [50800, 2800],
 81: [50800, 2400],
 82: [50800, 2000],
 83: [50800, 1600],
 84: [50800, 1200],
 85: [50800, 800],
 86: [50800, 400],
 87: [50800, 0],
 88: [51200, 2800],
 89: [51200, 2400],
 90: [51200, 2000],
 91: [51200, 1600],
 92: [51200, 1200],
 93: [51200, 800],
 94: [51200, 400],
 95: [51200, 0],
 96: [51600, 2800],
 97: [51600, 2400],
 98: [51600, 2000],
 99: [51600, 1600],
 100: [51600, 1200],
 101: [51600, 800],
 102: [51600, 400],
 103: [51600, 0],
 104: [52000, 2800],
 105: [52000, 2400],
 106: [52000, 2000],
 107: [52000, 1600],
 108: [52000, 1200],
 109: [52000, 800],
 110: [52000, 400],
 111: [52000, 0],
 112: [52400, 2800],
 113: [52400, 2400],
 114: [52400, 2000],
 115: [52400, 1600],
 116: [52400, 1200],
 117: [52400, 800],
 118: [52400, 400],
 119: [52400, 0],
 120: [52800, 2800],
 121: [52800, 2400],
 122: [52800, 2000],
 123: [52800, 1600],
 124: [52800, 1200],
 125: [52800, 800],
 126: [52800, 400],
 127: [52800, 0]}
```

I apologize in advance if this is due to some stupid error I made. I have thus far run out of ideas on what could be causing this and started to think there might be some undesired scaling of some parameter somewhere. Or that simply I am retrieving the spiketimes from the wrong place or in the wrong way.

Any help will be much appreciated :)

Thank you very much,
Aitor

**Comment by yger** [comment]
Actually, just to check that everything is fine, do you see similar observations before the merging the results? Because the fact that after merging, you have more spikes in a template than on one electrode is possible, since you are merging results of several templates (a template is now a combination of several others). And based on your parameters (auto_mode = 0.2), it may be than you are merging too much the templates. If you want, you can send me a zip file with just the results/.params files (the raw data may not be necessary such that I can double check everything)

**Comment by yger** [comment]
And also, a point that you are missing : clusters_0 only contains the spikes USED DURING THE CLUSTERING (so this is only a very small subset of all spikes that have been fitted). Having much more spikes in temp_0 than in the cluster_0 is thus completly normal. This is the whole point of template matching based algorithms : templates are obtained quickly on a subset of the spikes (saved for visualization if needed), and then spike times are assigned during the fitting procedure

**Comment by morales-gregorio** [comment]
Hey @yger !

Thanks a lot for your response!

Just for a quick check I see the same number of spikes in the template 0 before merging:
```
>>> import h5py
>>> result = h5py.File('NSP1.result.hdf5', mode='r')
>>> result['spiketimes']['temp_0']
<HDF5 dataset "temp_0": shape (19426,), type "<u4">
```
So I would not think that the merging step is where all the extra templates would come from. I was not aware that the clustering is applied to a subsample, that could explain the mismatch.

The sanity plot looks very reasonable to me:
![waveforms_neg_0](https://user-images.githubusercontent.com/43403140/72663613-e2e9d480-39f4-11ea-858a-58962ad2262c.png)
However, when I extract the waveforms from the filtered signal (same template of course) I get quite much noisier results.
![NSP_1_elec1_unit_0](https://user-images.githubusercontent.com/43403140/72663643-3bb96d00-39f5-11ea-9d52-966b7d028acb.png)

Very few of my extracted waveforms are not so noisy, for example:
![NSP_1_elec119_unit_2](https://user-images.githubusercontent.com/43403140/72663704-b3879780-39f5-11ea-887f-acecebaa46f6.png)

So this is why I have a feeling that the spiketimes in the results are being 'contaminated' with the spiketimes from nearby electrodes. 

I just realised that I applied thresholding from both sides (not just negative as the params I wrote in the issue, sorry about that). But that is the only difference with the used parameters (double checked).

I uploaded the spyking circus result files to a container [here](https://fz-juelich.sciebo.de/s/Jb5asK23CgSb566). Hope that can help to pinpoint the problem.

I really hope this problem is not too stupid and wasting your time. Thanks a lot for the help! :)

**Comment by yger** [comment]
Ok, so again, few points: the anity plot you are looking at is not made from the raw data, so you can not compare it to the snippets you are collecting manually. Let me explain : in the sanity plots, waveforms are reconstructed from the PCA (so they are smoothed, with some eneergy loss), and also the information at the border has been discarded (convolved with a Hanning window). This is to boost the clustering, since we really want it to focus on the core of the snippets In addition, these snippets, collecting during the clustering, are selected such that they should be "rather" isolated, i.e. not contaminated by nearby spikes. This is again to make sure that, since we are clustering only from a subsample of the spikes, we'll have the cleanest template (as the centroid of the cluster). The template is, of course, estimated from the raw data, and not as the mean/median over the curves you see in the sanity plot, for all the reasons explained below.
The waveforms you are showing me, from the filtered results, looks pretty clean to me. Keep in mind that what matters is only the center of the window, since at the borders, your template is 0 and thus you have only noise.

**Comment by morales-gregorio** [comment]
I understand, I was not fully aware of the methods then. Thanks a lot for your help!
